src/routing_table/prefix.rs

```diff
@@ -17,103 +17,222 @@
     /// is returned.
-    pub fn pushed(mut self, bit: bool) -> Prefix {
-        self.name = self.name.with_bit(self.bit_count(), bit);
-        self.bit_count = cmp::min(self.bit_count + 1, XOR_NAME_BITS as u16);
+    pub fn pushed(mut self, bit: bool) -> Self {
+        self.inner.name = self.inner.name.with_bit(self.bit_count(), bit);
+        self.inner.bit_count = cmp::min(self.inner.bit_count + 1, XOR_NAME_BITS as u16);
         self
     }
 
     /// Returns a prefix copying the first `bitcount() - 1` bits from `self`,
     /// or `self` if it is already empty.
-    pub fn popped(mut self) -> Prefix {
-        if self.bit_count > 0 {
-            self.bit_count -= 1;
+    pub fn popped(mut self) -> Self {
+        if self.inner.bit_count > 0 {
+            self.inner.bit_count -= 1;
             // unused bits should be zero:
-            self.name = self.name.with_bit(self.bit_count(), false);
+            self.inner.name = self.inner.name.with_bit(self.bit_count(), false);
         }
         self
     }
 
+    /// Returns the neighbouring prefix differing in the `i`-th bit
+    /// If `i` is larger than our bit count, `self` is returned
+    pub fn with_flipped_bit(&self, i: usize) -> Self {
+        if i >= self.bit_count() {
+            *self
+        } else {
+            Self::new(
+                self.bit_count(),
+                self.inner.name.with_flipped_bit(i),
+                self.version,
+            )
+        }
+    }
+
+    /// Returns the same prefix, with the last bit flipped, or unchanged, if empty.
+    pub fn sibling(&self) -> Self {
+        if self.bit_count() > 0 {
+            self.with_flipped_bit(self.bit_count() - 1)
+        } else {
+            *self
+        }
+    }
+
+    /// Returns whether the namespace defined by `self` is covered by prefixes in the `prefixes`
+    /// set
+    pub fn is_covered_by<'a, I>(&self, prefixes: I) -> bool
+    where
+        I: IntoIterator<Item = &'a Self> + Clone,
+    {
+        let max_prefix_len = prefixes
+            .clone()
+            .into_iter()
+            .map(|x| x.bit_count())
+            .max()
+            .unwrap_or(0);
+        self.is_covered_by_impl(prefixes, max_prefix_len)
+    }
+
+    fn is_covered_by_impl<'a, I>(&self, prefixes: I, max_prefix_len: usize) -> bool
+    where
+        I: IntoIterator<Item = &'a Self> + Clone,
+    {
+        prefixes.clone().into_iter().any(|x| {
+            x.is_compatible(self) && x.bit_count() <= self.bit_count()
+        }) ||
+            (self.bit_count() <= max_prefix_len &&
+                 self.pushed(false).is_covered_by_impl(
+                    prefixes.clone(),
+                    max_prefix_len,
+                ) &&
+                 self.pushed(true).is_covered_by_impl(
+                    prefixes,
+                    max_prefix_len,
+                ))
+    }
+}
+
+impl Deref for Prefix {
```

Comment:
This is a clever move, but I wonder if it may make the calling code more obvious if we just take the more tedious approach of implementing the handful of missing methods in `Prefix`, e.g.
```rust
pub fn is_neighbour(&self, other: &Self) -> bool {
    self.inner.is_neighbour(&other.inner)
}
```

Reply:
Yeah, I went back and forth on this. Not really decided. With `Deref`, you can also pass `Prefix` to functions that take `&UnversionedPrefix` which might be useful or confusing, depends on how you look at it.

Reply:
> might be useful or confusing

Was both for me initially 😄   I think because of the risk of how badly things could go if we accidentally don't consider versions when we really should, I'd go for the more explicit route, e.g. force the user to call `.unversioned()`.  However, it's probably worth getting a few more opinions before you make the change?

Reply:
I will prefer a more explicit way, i.e. calling `.unversioned()`, as the code will be easier to read.